file.c

```diff
@@ -0,0 +1,289 @@
+    f->olen = 0;
+    errno = 0;
+}
+
+size_t fileread(mfile f, void *buf, size_t len, int64_t deadline) {
+    /* If there's enough data in the buffer it's easy. */
+    if(f->ilen >= len) {
+        memcpy(buf, &f->ibuf[f->ifirst], len);
+        f->ifirst += len;
+        f->ilen -= len;
+        errno = 0;
+        return len;
+    }
+
+    /* Let's move all the data from the buffer first. */
+    char *pos = (char*)buf;
+    size_t remaining = len;
+    memcpy(pos, &f->ibuf[f->ifirst], f->ilen);
+    pos += f->ilen;
+    remaining -= f->ilen;
+    f->ifirst = 0;
+    f->ilen = 0;
+
+    mill_assert(remaining);
+    while(1) {
+        if(remaining > MILL_FILE_BUFLEN) {
+            /* If we still have a lot to read try to read it in one go directly
+             into the destination buffer. */
+            ssize_t sz = read(f->fd, pos, remaining);
+            if(!sz) {
+                return len - remaining;
+            }
+            if(sz == -1) {
+                if(errno != EAGAIN && errno != EWOULDBLOCK)
+                    return len - remaining;
+                sz = 0;
+            }
+            if((size_t)sz == remaining) {
+                errno = 0;
+                return len;
+            }
+            pos += sz;
+            remaining -= sz;
+            if (sz != 0 && fileeof(f)) {
+                return len - remaining;
+            }
+        }
+        else {
+            /* If we have just a little to read try to read the full connection
+             buffer to minimise the number of system calls. */
+            ssize_t sz = read(f->fd, f->ibuf, MILL_FILE_BUFLEN);
+            if(!sz) {
+                return len - remaining;
+            }
+            if(sz == -1) {
+                if(errno != EAGAIN && errno != EWOULDBLOCK)
+                    return len - remaining;
+                sz = 0;
+            }
+            if((size_t)sz < remaining) {
+                memcpy(pos, f->ibuf, sz);
+                pos += sz;
+                remaining -= sz;
+                f->ifirst = 0;
+                f->ilen = 0;
+            }
+            else {
+                memcpy(pos, f->ibuf, remaining);
+                f->ifirst = remaining;
+                f->ilen = sz - remaining;
+                errno = 0;
+                return len;
+            }
+            if (sz != 0 && fileeof(f)) {
+                return len - remaining;
+            }
+        }
+
+        /* Wait till there's more data to read. */
+        int res = fdwait(f->fd, FDW_IN, deadline);
+        if (!res) {
+            errno = ETIMEDOUT;
+            return len - remaining;
+        }
+    }
+}
+
+void fileclose(mfile f) {
+    fdclean(f->fd);
+    int rc = close(f->fd);
+    mill_assert(rc == 0);
+    free(f);
+    return;
+}
+
+mfile fileattach(int fd) {
+    struct mill_file *f = malloc(sizeof(struct mill_file));
+    if(!f) {
+        errno = ENOMEM;
+        return NULL;
+    }
+    int nfd = dup(fd);
+    mill_filetune(nfd);
```

Comment:
I think the root of the problem is not about duplication, but about making shared file descriptors non blocking. Removing the attach function would help in that sense, but the global wrappers of the streams would have to make them nonblocking so the problem would still be there. I truly think this is an issue with POSIX itself so I guess it would be ok to leave the `attach` and make `tune` public.